### PR TITLE
DP-229 - BE- Share my information - Generating code (API)

### PR DESCRIPTION
**Pull request by maciej-goaco**

### CODE REVIEW

File: Services/CO.CDP.DataSharing.WebApi.Tests/CO.CDP.DataSharing.WebApi.Tests.csproj [hunk 1]
```diff
@@ -11,7 +11,9 @@
     </PropertyGroup>
 
     <ItemGroup>
+        <PackageReference Include="FluentAssertions" Version="6.12.0" />
         <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.10.0" />
+        <PackageReference Include="Nanoid" Version="3.1.0" />
```
Comment (jakzal) [inline Services/CO.CDP.DataSharing.WebApi.Tests/CO.CDP.DataSharing.WebApi.Tests.csproj:16]: This is already in CO.CDP.DataSharing.WebApi.csproj, so shouldn't be needed here.

**Review by dbgoaco — APPROVED** [review]

File: Services/CO.CDP.DataSharing.WebApi.Tests/CO.CDP.DataSharing.WebApi.Tests.csproj (hunk 1 above)
Comment (maciej-goaco) [reply Services/CO.CDP.DataSharing.WebApi.Tests/CO.CDP.DataSharing.WebApi.Tests.csproj]: done.

File: Services/CO.CDP.DataSharing.WebApi/Api/DataSharing.cs [hunk 2]
```diff
@@ -174,24 +176,18 @@ public static void UseDataSharingEndpoints(this WebApplication app)
...
-        app.MapPost("/share/data", (ShareRequest shareRequest) => Results.Ok(new ShareReceipt
-        {
-            ShareCode = Guid.NewGuid().ToString(),
-            ExpiresAt = shareRequest.ExpiresAt,
-            FormId = shareRequest.SupplierFormId,
-            FormVersionId = "20240317",
-            Permissions = shareRequest.Permissions
-        }
-            ))
+        app.MapPost("/share/data", async (ShareRequest shareRequest, IUseCase<ShareRequest, ShareReceipt?> useCase) =>
```
Comment (jakzal) [inline Services/CO.CDP.DataSharing.WebApi/Api/DataSharing.cs:179]: Should be `IUseCase<ShareRequest, ShareReceipt>`

File: Services/CO.CDP.DataSharing.WebApi/Extensions/ServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
+using CO.CDP.DataSharing.WebApi.Model;
+
+namespace CO.CDP.DataSharing.WebApi.Extensions;
+
+public static class ServiceCollectionExtensions
```
Comment (jakzal) [inline Services/CO.CDP.DataSharing.WebApi/Extensions/ServiceCollectionExtensions.cs:5]: We usually add a basic test case for this.

File: Services/CO.CDP.DataSharing.WebApi/UseCase/GenerateShareCodeUseCase.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+using AutoMapper;
+using CO.CDP.DataSharing.WebApi.Extensions;
+using CO.CDP.DataSharing.WebApi.Model;
+using CO.CDP.OrganisationInformation.Persistence;
+using CO.CDP.OrganisationInformation.Persistence.Forms;
+
+namespace CO.CDP.DataSharing.WebApi.UseCase;
+
+public class GenerateShareCodeUseCase(IFormRepository formRepository, IMapper mapper)
```
Comment (jakzal) [inline Services/CO.CDP.DataSharing.WebApi/UseCase/GenerateShareCodeUseCase.cs:9]: Rename the test case as well please.

File: Services/CO.CDP.DataSharing.WebApi.Tests/AutoMapper/AutoMapperFixture.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+using AutoMapper;
+using CO.CDP.DataSharing.WebApi.AutoMapper;
+
+namespace CO.CDP.DataSharing.Tests.AutoMapper;
+
+public class AutoMapperFixture
```
Comment (jakzal) [inline Services/CO.CDP.DataSharing.WebApi.Tests/AutoMapper/AutoMapperFixture.cs:6]: This fixture is meant to be used with a test case. The test case is missing. Have a look at other projects. The test case ensures that automapper config can be loaded.

https://github.com/cabinetoffice/GCGS-Central-Digital-Platform/blob/a31c9205266b21418aa39449f1ecbc49cc58afab/Services/CO.CDP.Tenant.WebApi.Tests/AutoMapper/WebApiToPersistenceProfileTest.cs#L3-L10

File: Services/CO.CDP.DataSharing.WebApi/Api/DataSharing.cs (hunk 2 above)
Comment (maciej-goaco) [reply Services/CO.CDP.DataSharing.WebApi/Api/DataSharing.cs:179]: done.

File: Services/CO.CDP.DataSharing.WebApi.Tests/AutoMapper/AutoMapperFixture.cs (hunk 5 above)
Comment (maciej-goaco) [reply Services/CO.CDP.DataSharing.WebApi.Tests/AutoMapper/AutoMapperFixture.cs:6]: done.

File: Services/CO.CDP.DataSharing.WebApi.Tests/UseCase/GenerateShareCodeUseCaseTest.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            SubmissionState = SubmissionState.Submitted,
+            SubmittedAt = DateTime.UtcNow,
+            FormVersionId = formVersionId,
+            BookingReference = string.Empty
+        };
+
+        _repository.Setup(r => r.GetSharedConsentAsync(shareRequest.FormId, shareRequest.OrganisationId)).ReturnsAsync(sharedConsent);
+
+        var found = await UseCase.Execute(shareRequest);
+        var expectedShareReceipt = new ShareReceipt()
```
Comment (jakzal) [inline Services/CO.CDP.DataSharing.WebApi.Tests/UseCase/GenerateShareCodeUseCaseTest.cs:77]: seems unused.

File: Services/CO.CDP.DataSharing.WebApi/Api/DataSharing.cs [hunk 7]
```diff
@@ -174,24 +176,18 @@ public static void UseDataSharingEndpoints(this WebApplication app)
...
-            ShareCode = Guid.NewGuid().ToString(),
-            ExpiresAt = shareRequest.ExpiresAt,
-            FormId = shareRequest.SupplierFormId,
-            FormVersionId = "20240317",
-            Permissions = shareRequest.Permissions
-        }
-            ))
+        app.MapPost("/share/data", async (ShareRequest shareRequest, IUseCase<ShareRequest, ShareReceipt> useCase) =>
+                await useCase.Execute(shareRequest)
+                    .AndThen(supplierInfo => Results.Ok(supplierInfo)))
```
Comment (jakzal) [inline Services/CO.CDP.DataSharing.WebApi/Api/DataSharing.cs:181]: not a `supplierInfo`,  but a `shareReceipt`.

File: Services/CO.CDP.OrganisationInformation.Persistence/DatabaseFormRepository.cs [hunk 8]
```diff
@@ -33,6 +40,16 @@ public async Task SaveFormAsync(Form form)
 
     #endregion
 
+    #region Shared Consents Methods
+
+    public async Task<SharedConsent?> GetSharedConsentAsync(Guid formId, Guid organisationId)
+    {
+        return await context.Set<SharedConsent>()
+            .FirstOrDefaultAsync(s => s.Form.Guid == formId && s.Organisation.Guid == organisationId);
```
Comment (jakzal) [inline Services/CO.CDP.OrganisationInformation.Persistence/DatabaseFormRepository.cs:48]: Actually, we should find the consent that has no shared code assigned. There should only ever be one.

File: Services/CO.CDP.DataSharing.WebApi.Tests/UseCase/GenerateShareCodeUseCaseTest.cs (hunk 6 above)
Comment (maciej-goaco) [reply Services/CO.CDP.DataSharing.WebApi.Tests/UseCase/GenerateShareCodeUseCaseTest.cs:77]: done.

File: Services/CO.CDP.DataSharing.WebApi/Api/DataSharing.cs (hunk 7 above)
Comment (maciej-goaco) [reply Services/CO.CDP.DataSharing.WebApi/Api/DataSharing.cs:181]: done.

File: Services/CO.CDP.OrganisationInformation.Persistence/DatabaseFormRepository.cs (hunk 8 above)
Comment (maciej-goaco) [reply Services/CO.CDP.OrganisationInformation.Persistence/DatabaseFormRepository.cs:48]: done.

**Review by jakzal — APPROVED** [review]

**Pull request merged**